virtcontainers/device/manager/manager.go

```diff
@@ -21,59 +26,125 @@ const (
+	// in case of ID collision
+	if devInfo.ID, err = dm.newDeviceID(); err != nil {
+		return nil, err
+	}
 	if isVFIO(path) {
-		return drivers.NewVFIODevice(devInfo), nil
+		return drivers.NewVFIODevice(dm, &devInfo), nil
 	} else if isBlock(devInfo) {
 		if devInfo.DriverOptions == nil {
 			devInfo.DriverOptions = make(map[string]string)
 		}
 		devInfo.DriverOptions["block-driver"] = dm.blockDriver
-		return drivers.NewBlockDevice(devInfo), nil
+		return drivers.NewBlockDevice(dm, &devInfo), nil
 	} else {
 		deviceLogger().WithField("device", path).Info("Device has not been passed to the container")
-		return drivers.NewGenericDevice(devInfo), nil
+		return drivers.NewGenericDevice(dm, &devInfo), nil
 	}
 }
 
-// NewDevices creates bundles of devices based on array of DeviceInfo
-func (dm *deviceManager) NewDevices(devInfos []config.DeviceInfo) ([]api.Device, error) {
+// NewDevice creates bundles of devices based on array of DeviceInfo
+func (dm *deviceManager) NewDevice(devInfo config.DeviceInfo) (api.Device, error) {
+	dm.Lock()
+	defer dm.Unlock()
+	dev, err := dm.createDevice(devInfo)
+	if err == nil {
+		dm.devices[dev.DeviceID()] = dev
+	}
+	return dev, err
+}
+
+// DeviceAttached returns a device list of all hotplugged devices with same hostpath
+// with input paramater device
+func (dm *deviceManager) AttachedDevices(device api.Device) []api.Device {
 	var devices []api.Device
+	devInfo := device.GetDeviceInfo()
+	for _, dev := range dm.devices {
+		info := dev.GetDeviceInfo()
+		// regard two devices as same if they has same HostPath and DevType
+		if info.Hotplugged && info.HostPath == devInfo.HostPath &&
+			info.DevType == devInfo.DevType {
+			devices = append(devices, dev)
+		}
+	}
+	return devices
+}
 
-	for _, devInfo := range devInfos {
-		device, err := dm.createDevice(devInfo)
+func (dm *deviceManager) newDeviceID() (string, error) {
+	for i := 0; i < 5; i++ {
+		// generate an random ID
+		randBytes, err := utils.GenerateRandomBytes(8)
 		if err != nil {
-			return nil, err
+			return "", err
 		}
-		devices = append(devices, device)
-	}
+		id := hex.EncodeToString(randBytes)
 
-	return devices, nil
+		// check ID collision, choose another one if ID is in use
+		if _, ok := dm.devices[id]; !ok {
+			return id, nil
+		}
+	}
+	return "", ErrIDExhausted
 }
 
-// NewDeviceManager creates a deviceManager object behaved as api.DeviceManager
-func NewDeviceManager(blockDriver string) api.DeviceManager {
-	dm := &deviceManager{}
-	if blockDriver == VirtioBlock {
-		dm.blockDriver = VirtioBlock
-	} else {
-		dm.blockDriver = VirtioSCSI
-	}
+func (dm *deviceManager) Populate(dev api.Device) api.Device {
+	dm.Lock()
+	defer dm.Unlock()
 
-	return dm
+	switch d := dev.(type) {
+	case *drivers.BlockDevice:
+		d.DevManager = dm
+		dm.devices[d.DeviceID()] = d
+		return d
+	case *drivers.VFIODevice:
+		d.DevManager = dm
+		dm.devices[d.DeviceID()] = d
+		return d
+	case *drivers.GenericDevice:
```

Comment:
As these three cases are identical, you could add explicit `fallthrough`'s. Or something like:

```
switch d := dev.(type) {
	case *drivers.BlockDevice:
	case *drivers.VFIODevice:
	case *drivers.GenericDevice:
	default:
		return nil
}

d.DevManager = dm
dm.devices[d.DeviceID()] = d
return d
```

Reply:
`fallthrough` isn't supported in type switch in my test. But latter one looks good, will modify, thank!

Reply:
Oops, sorry, just found that the latter one doesn't work either.
I need to use `d` in type switch, if I put the logic at below, it can't visit d anymore.
So have to keep the original way.